server/src/main/java/org/elasticsearch/common/util/set/Sets.java

```diff
@@ -88,9 +76,13 @@ public static <T> boolean haveNonEmptyIntersection(Set<T> left, Set<T> right) {
      * @return the relative complement of the left set with respect to the right set
      */
     public static <T> Set<T> difference(Set<T> left, Set<T> right) {
-        Objects.requireNonNull(left);
-        Objects.requireNonNull(right);
-        return left.stream().filter(k -> right.contains(k) == false).collect(Collectors.toSet());
+        Set<T> set = new HashSet<>();
```

Comment:
Is it worth pre-sizing this to `left.size()` to avoid possible allocations (at the cost of more memory usage when the difference is small)? Or maybe meet-in-the-middle at size/2?

Reply:
Hmm it's hard to make a good guess about the result set size here as you point out. What will happen though is that the method gets bigger if we do size magic with it, to me it seems like just going for the default might be the fastest option on average but it's hard to prove that. Though it certainly got me what I wanted in the one spot where this is hot :)

Reply:
I suppose you can just do this manually if there is a point where it does actually matter